---
title: "BiocPBG: application of PyTorch biggraph in single cell multiomics"
author: "Vincent J. Carey, stvjc at channing.harvard.edu"
date: "`r format(Sys.time(), '%B %d, %Y')`"
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{BiocPBG: application of PyTorch biggraph in single cell multiomics}
  %\VignetteEncoding{UTF-8}
output:
  BiocStyle::html_document:
    highlight: pygments
    number_sections: yes
    theme: united
    toc: yes
---

# Introduction

Inspired by the SIMBA algorithm of the Pinello lab ([paper](https://www.nature.com/articles/s41592-023-01899-8), [github](https://github.com/pinellolab/simba)),
this package illustrates the use of embeddings
produced with PyTorch-BigGraph
([paper](https://mlsys.org/Conferences/2019/doc/2019/71.pdf),
[github](https://github.com/facebookresearch/PyTorch-BigGraph))
in the analysis of single-cell experiments.

## Prerequisites

### Python 3.10

There is a bug in the base PyTorch-BigGraph distribution
from facebook research.  A 'patched' version is available
at `https://github.com/vjcitn/PyTorch-BigGraph`.  Clone
that repo and use branch `vjc_idx_fix_gpu`.  The use
`pip install -e .` in the repo root folder with the version
of pip that installs to the locations employed by
reticulate.  Set `RETICULATE_PYTHON` if necessary to
ensure that the right version of python is used.

### R 4.3+

To use this package with R 4.3 or later, be sure that
the R packages devtools, remotes, reticulate, and BiocManager
are available.
Then the package can be installed with
```
BiocManager::install("vjcitn/BiocPBG")
```

# RNA-seq example

## Building the embeddings

The SIMBA tutorials include a [jupyter notebook](https://simba-bio.readthedocs.io/en/latest/rna_10xpmbc_all_genes_v1.2.html)
working through the embedding of RNA-seq 
results on 2700 PBMCs.  We used a GPU-enabled
machine in the NSF ACCESS Jetstream2 system
(allocation BIR190004) to create 50-dimensional
embeddings of cells and genes with a
[script](https://github.com/vjcitn/BiocPBG/blob/main/inst/scripts/sce_to_emb50.R).  The primary outcomes are in a stored R list called `nn50`.

```{r lknn50, message=FALSE}
library(BiocPBG)
data(nn50)
str(nn50)
```
The `config` element can be useful in interactive computing
in a given R/reticulate session.  After serialization, the
element is a null pointer.

See the appendix for additional details on the formation
of the input data, which involves discretization of
normalized expression measures.

## Informal assessment of cell embedding

When the embedding outputs are reduced using UMAP with
default settings, we have:

```{r doum,message=FALSE}
library(uwot)
library(plotly)
data(t3k)
set.seed(1234) # reproducible umap
cu = umap(t(nn50$cemb))
cud = data.frame(um1=cu[,1], um2=cu[,2], barcode=nn50$cents)
colnames(t3k) = t3k$Barcode # need to set barcode
t3kb = t3k[, cud$barcode]   # reorder columns
cud$celltype = t3kb$celltype2 #HPCA
plot_ly(data=cud, x=~um1, y=~um2, color=~celltype, text=~celltype, alpha=.6)
```

The labeling here is based on the `HumanPrimaryCellAtlasData`
element of [celldex](https://bioconductor.org/packages/celldex),
applied using the [SingleR](https://bioconductor.org/packages/SingleR)
method.
