---
title: "BiocPBG: application of PyTorch biggraph in single cell multiomics"
author: "Vincent J. Carey, stvjc at channing.harvard.edu"
date: "`r format(Sys.time(), '%B %d, %Y')`"
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{BiocPBG: application of PyTorch biggraph in single cell multiomics}
  %\VignetteEncoding{UTF-8}
output:
  BiocStyle::html_document:
    highlight: pygments
    number_sections: yes
    theme: united
    toc: yes
---

# Introduction

Inspired by the SIMBA algorithm of the Pinello lab ([paper](https://www.nature.com/articles/s41592-023-01899-8), [github](https://github.com/pinellolab/simba)),
this package illustrates the use of embeddings
produced with PyTorch-BigGraph
([paper](https://mlsys.org/Conferences/2019/doc/2019/71.pdf),
[github](https://github.com/facebookresearch/PyTorch-BigGraph))
in the analysis of single-cell experiments.

## Prerequisites

### Python 3.10

There is a bug in the base PyTorch-BigGraph distribution
from facebook research.  A 'patched' version is available
at `https://github.com/vjcitn/PyTorch-BigGraph`.  Clone
that repo and use branch `vjc_idx_fix_gpu`.  The use
`pip install -e .` in the repo root folder with the version
of pip that installs to the locations employed by
reticulate.  Set `RETICULATE_PYTHON` if necessary to
ensure that the right version of python is used.

### R 4.3+

To use this package with R 4.3 or later, be sure that
the R packages devtools, remotes, reticulate, and BiocManager
are available.
Then the package can be installed with
```
BiocManager::install("vjcitn/BiocPBG")
```

# RNA-seq example

## Building the embeddings

The SIMBA tutorials include a [jupyter notebook](https://simba-bio.readthedocs.io/en/latest/rna_10xpmbc_all_genes_v1.2.html)
working through the embedding of RNA-seq 
results on 2700 PBMCs.  We used a GPU-enabled
machine in the NSF ACCESS Jetstream2 system
(allocation BIR190004) to create 50-dimensional
embeddings of cells and genes with a
[script](https://github.com/vjcitn/BiocPBG/blob/main/inst/scripts/sce_to_emb50.R).  The primary outcomes are in a stored R list called `nn50`.

```{r lknn50, message=FALSE}
library(BiocPBG)
data(nn50)
str(nn50)
```
The `config` element can be useful in interactive computing
in a given R/reticulate session.  After serialization, the
element is a null pointer.

See the appendix for additional details on the formation
of the input data, which involves discretization of
normalized expression measures.

## Loss function trace

```{r lkloss}
plot(y=nn50$stats$stats.metrics.loss, 
  x=nn50$stats$epoch_idx, xlab="Epoch", ylab="Loss")
```

## Informal assessment of cell embedding

When the embedding outputs are reduced using UMAP with
default settings, we have the following
interactive display.

```{r doum,message=FALSE}
library(uwot)
library(plotly)
data(t3k)
set.seed(1234) # reproducible umap
cu = umap(t(nn50$cemb))
cud = data.frame(um1=cu[,1], um2=cu[,2], barcode=nn50$cents)
colnames(t3k) = t3k$Barcode # need to set barcode
t3kb = t3k[, cud$barcode]   # reorder columns
cud$celltype = t3kb$celltype2 #HPCA
plot_ly(data=cud, x=~um1, y=~um2, color=~celltype, text=~celltype, alpha=.6)
```

The labeling here is based on the `HumanPrimaryCellAtlasData`
element of [celldex](https://bioconductor.org/packages/celldex),
applied using the [SingleR](https://bioconductor.org/packages/SingleR)
method.

A similar display can be based on approximate principal components
based on the BigGraph embeddings:
```{r doapprox,message=FALSE}
library(irlba)
cp = prcomp_irlba(t(nn50$cemb), n=10)
cud = data.frame(pc1=cp$x[,1], pc2=cp$x[,2], 
  pc3=cp$x[,3], pc4=cp$x[,4], barcode=nn50$cents)
colnames(t3k) = t3k$Barcode # need to set barcode
t3kb = t3k[, cud$barcode]   # reorder columns
cud$celltype = t3kb$celltype # original
pairs(data.matrix(cud[,1:4]), col=factor(cud$celltype), pch=19)
```

Let's drill in on the PC1 x PC4 display:
```{r cont}
plot_ly(data=cud, y = ~pc1, x = ~pc4, color=~celltype)
```

PCA based on the normalized counts:
```{r lkrawpca}
cp = prcomp_irlba(t(assay(t3k, "logcounts")), n=10)
cud = data.frame(pc1=cp$x[,1], pc2=cp$x[,2], 
  pc3=cp$x[,3], pc4=cp$x[,4], celltype=t3k$celltype)
pairs(data.matrix(cud[,1:3]), col=factor(cud$celltype), pch=19)
```

## Gene barcodes

The estimated probabilities linking
genes to cell types are ordered to produce
barcode displays.

```{r showbar, message=FALSE, fig.width=unit(14, "cm"), fig.height=unit(10,"cm")}
data(nn50)
data(t3k)
t3k = add_simba_scores(nn50, t3k)
gns = c("CST3", "NKG7", "GAPDH", "MS4A1")
alldf = lapply(gns, function(x) simba_barplot_df(t3k, x,
  colvars = "celltype"))
alldf = do.call(rbind, alldf)
ggplot(alldf, aes(x=rank,xend=rank,y=0, yend=prob,colour=celltype)) + 
  geom_point() +
  geom_segment() + facet_wrap(~symbol, ncol=2) +
    theme(strip.text=element_text(size=24),
      legend.text=element_text(size=24),
      axis.text=element_text(size=24)) +
    guides(colour = guide_legend(override.aes =
       list(linetype=1, lwd=2)))
```

## Cell-type marker assessment via Gini index

At this time our implementation of the 'max' metric
is not faithful to the simba modules.  But the following code chunk comes
close to producing Figure 2e of the SIMBA paper.

```{r lkgin}
# t3k = add_simba_scores(nn50, t3k) # already done
gi = gini(t(assay(altExps(t3k)$sprobs)))
ma = as.numeric(max_scores(t3k))
mydf = data.frame(max=as.numeric(ma), gini=gi, gene=rowData(t3k)$Symbol)
litd = mydf[order(mydf$gini,decreasing=TRUE),][1:25,]
hh = mydf[mydf$gene=="GAPDH",] 
litd = rbind(litd, hh)
mm = ggplot(mydf, aes(x=max, y=gini, text=gene)) + geom_point() 
mm + ggrepel::geom_text_repel(data=litd, 
          aes(x=max, y=gini, label=gene), colour="purple")
```

