<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ BiocPBG</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">BiocPBG</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.10</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/BiocPBG.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/vjcitn/BiocPBG/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="add_simba_scores.html">add_simba_scores()</a></code> </p>
        </td>
        <td><p>update an SCE with simba edge scores</p></td>
      </tr><tr><td>
          <p><code><a href="disc_matrix.html">disc_matrix()</a></code> </p>
        </td>
        <td><p>discretize a dense matrix</p></td>
      </tr><tr><td>
          <p><code><a href="docol.html">docol()</a></code> </p>
        </td>
        <td><p>helper function</p></td>
      </tr><tr><td>
          <p><code><a href="embs_to_scores.html">embs_to_scores()</a></code> </p>
        </td>
        <td><p>Produce a SummarizedExperiment instance with assay elements
(i,j) the simba score relating gene i to cell j</p></td>
      </tr><tr><td>
          <p><code><a href="fb15k_folder.html">fb15k_folder()</a></code> </p>
        </td>
        <td><p>set up folder with fb15k demo data</p></td>
      </tr><tr><td>
          <p><code><a href="flatten_json.html">flatten_json()</a></code> </p>
        </td>
        <td><p>Helper function to recursively flatten nested lists
produced by Chat-GPT 3</p></td>
      </tr><tr><td>
          <p><code><a href="gini.html">gini()</a></code> </p>
        </td>
        <td><p>gini coefficients from rows of a matrix</p></td>
      </tr><tr><td>
          <p><code><a href="ingest_training_stats.html">ingest_training_stats()</a></code> </p>
        </td>
        <td><p>get pbg_metrics data frames</p></td>
      </tr><tr><td>
          <p><code><a href="make_entity_schema.html">make_entity_schema()</a></code> </p>
        </td>
        <td><p>produce EntitySchema</p></td>
      </tr><tr><td>
          <p><code><a href="make_rel_schema.html">make_rel_schema()</a></code> </p>
        </td>
        <td><p>produce RelationSchema for torchbiggraph configuration</p></td>
      </tr><tr><td>
          <p><code><a href="make_triples.html">make_triples()</a></code> </p>
        </td>
        <td><p>make triples from a discretized SCE</p></td>
      </tr><tr><td>
          <p><code><a href="make_trvate.html">make_trvate()</a></code> </p>
        </td>
        <td><p>convert data.frame to train/test/validate subsamples</p></td>
      </tr><tr><td>
          <p><code><a href="max_scores.html">max_scores()</a></code> </p>
        </td>
        <td><p>obtain the simba max metric for genes</p></td>
      </tr><tr><td>
          <p><code><a href="nn50.html">nn50</a></code> </p>
        </td>
        <td><p>embeddings for PBMC3K based on 50 epochs of training by
torch-biggraph for 4780 highly variable genes on 2700 cells.</p></td>
      </tr><tr><td>
          <p><code><a href="norm_scores.html">norm_scores()</a></code> </p>
        </td>
        <td><p>obtain the simba norm metric for genes</p></td>
      </tr><tr><td>
          <p><code><a href="path_to_15k_tgz.html">path_to_15k_tgz()</a></code> </p>
        </td>
        <td><p>cache and/or retrieve path to fb15k.tgz</p></td>
      </tr><tr><td>
          <p><code><a href="pca_CG.html">pca_CG()</a></code> </p>
        </td>
        <td><p>update a SingleCellExperiment with PCA (via irlba) reductions of 
embeddings produced with PyTorch-BigGraph</p></td>
      </tr><tr><td>
          <p><code><a href="sce_to_embeddings.html">sce_to_embeddings()</a></code> </p>
        </td>
        <td><p>produce torchbiggraph embeddings for discretized single-cell
RNA-seq measures in a SingleCellExperiment</p></td>
      </tr><tr><td>
          <p><code><a href="sce_to_triples.html">sce_to_triples()</a></code> </p>
        </td>
        <td><p>filter a discretized SCE using getTopHVGs from scran, the produce triples (edges)
(cell - weight - gene)</p></td>
      </tr><tr><td>
          <p><code><a href="setup_config_schema.html">setup_config_schema()</a></code> </p>
        </td>
        <td><p>make ConfigSchema</p></td>
      </tr><tr><td>
          <p><code><a href="simba_barplot_df.html">simba_barplot_df()</a></code> </p>
        </td>
        <td><p>set up a data.frame for barplot visualization for a gene</p></td>
      </tr><tr><td>
          <p><code><a href="t3k.html">t3k</a></code> </p>
        </td>
        <td><p>a SingleCellExperiment for PBMC3K that includes assay elements
logcounts (produced by scran::logNormCounts) and disc (produced
by BiocPBG::disc_matrix)</p></td>
      </tr><tr><td>
          <p><code><a href="train_eval.html">train_eval()</a></code> </p>
        </td>
        <td><p>run training and (optionally) evaluation</p></td>
      </tr><tr><td>
          <p><code><a href="triples_to_hdf5.html">triples_to_hdf5()</a></code> </p>
        </td>
        <td><p>use config object</p></td>
      </tr><tr><td>
          <p><code><a href="triples_to_hdf5_raw.html">triples_to_hdf5_raw()</a></code> </p>
        </td>
        <td><p>produce HDF5 files for edges defined in tsv files of triples (left, rel, right)</p></td>
      </tr><tr><td>
          <p><code><a href="umap_CG.html">umap_CG()</a></code> </p>
        </td>
        <td><p>update a SingleCellExperiment with UMAP reductions of 
embeddings produced with PyTorch-BigGraph</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Vincent Carey.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

